src/core.jl

```diff
@@ -87,101 +235,161 @@ function issubset{T<:SimpleGraph}(g::T, h::T)
+    g.fm.m += 1
+    g.fm.n += 1
+    push!(g.fm.colptr, g.fm.colptr[end])
+    return g.fm.m
+end
 
-    return n
+function add_vertex!(g::SparseDiGraph)
+    g.fm.m += 1
+    g.fm.n += 1
+    push!(g.fm.colptr, g.fm.colptr[end])
+    g.bm.m += 1
+    g.bm.n += 1
+    push!(g.bm.colptr, g.bm.colptr[end])
+    return g.fm.m
 end
 
+
 """Add `n` new vertices to the graph `g`."""
-function add_vertices!(g::SimpleGraph, n::Integer)
-    for i = 1:n
-        add_vertex!(g)
-    end
-    return nv(g)
+function add_vertices!(g::SparseGraph, n::Integer)
+    g.fm.m += n
+    g.fm.n += n
+    append!(g.fm.colptr, fill(g.fm.colptr[end], n))
+    return g.fm.m
+end
+
+function add_vertices!(g::SparseDiGraph, n::Integer)
+    g.fm.m += n
+    g.fm.n += n
+    append!(g.fm.colptr, fill(g.fm.colptr[end], n))
+    g.bm.m += n
+    g.bm.n += n
+    append!(g.bm.colptr, fill(g.bm.colptr[end], n))
+    return g.fm.m
 end
 
 """Return true if the graph `g` has an edge from `src` to `dst`."""
-has_edge(g::SimpleGraph, src::Int, dst::Int) = has_edge(g,Edge(src,dst))
+has_edge(g::SimpleGraph, e::Edge) = has_edge(g, src(e), dst(e))
+has_edge(g::SimpleGraph, s::Int, d::Int) = d <= nv(g) && s <= nv(g) && fmat(g)[d, s]
 
 """Return an Array of the edges in `g` that arrive at vertex `v`."""
 in_edges(g::SimpleGraph, v::Int) = [Edge(x,v) for x in badj(g,v)]
 """Return an Array of the edges in `g` that emanate from vertex `v`."""
 out_edges(g::SimpleGraph, v::Int) = [Edge(v,x) for x in fadj(g,v)]
 
+
 """Return true if `v` is a vertex of `g`."""
 has_vertex(g::SimpleGraph, v::Int) = v in vertices(g)
 
 """The number of vertices in `g`."""
-nv(g::SimpleGraph) = length(vertices(g))
+nv(g::AbstractSparseGraph) = fmat(g).m
+
+
 """The number of edges in `g`."""
 ne(g::SimpleGraph) = length(edges(g))
 
-"""Add a new edge to `g` from `src` to `dst`.
-
-Note: An exception will be raised if the edge is already in the graph
-or if the vertex is not contained in the graph.
+doc"""Density is defined as the ratio of the number of actual edges to the
+number of possible edges. This is $|v| |v-1|$ for directed graphs and
+$(|v| |v-1|) / 2$ for undirected graphs.
 """
-function add_edge!(g::SimpleGraph, e::Edge)
-    has_edge(g,e) && error("Edge $e already in graph")
-    (has_vertex(g,src(e)) && has_vertex(g,dst(e))) || throw(BoundsError())
-    unsafe_add_edge!(g,e)
+density(g::SparseGraph) = (2*ne(g)) / (nv(g) * (nv(g)-1))
+density(g::SparseDiGraph) = ne(g) / (nv(g) * (nv(g)-1))
+
+
+"""Add a new edge to `g` from `src` to `dst`."""
+add_edge!(g::SimpleGraph, s::Int, d::Int) = add_edge!(g, Edge(s,d))
+
+function add_edge!(g::SparseGraph, e::Edge)
+    s, d = src(e), dst(e)
+    g.fm[d,s] = true
+    g.fm[s,d] = true
+    push!(g.edges, e)
+    return e
 end
 
-add_edge!(g::SimpleGraph, src::Int, dst::Int) = add_edge!(g, Edge(src,dst))
+function add_edge!(g::SparseDiGraph, e::Edge)
+    s, d = src(e), dst(e)
+    g.fm[d,s] = true
+    g.bm[s,d] = true
```

Comment:
If you're tracking both forward and backward adjacencies, how else could you do it? Open to suggestions.